Introducing Launcher::overcommit, improving CI formatting

riesentoaster opened the pull request:
As discussed in #2665.

This is a very simple implementation that just spawns n clients per core. I'm not sure how this interacts with fixing a certain client to a core (via core affinity of the thread/process).

I've also added some cleanup (pre-commit scripts that have been complaining for a bit, unnecessary cfg restrictions).

rmalmain commented:
looks useful.
wouldn't it make more sense to implement this as a wrapper executor instead?
if it's implemented as a launcher, we cannot reuse this easily with different flavors like centalized

riesentoaster commented:
Also: I ran `scripts/fmt_all.sh` on MacOS. It suggested the changes in [fuzzers/inprocess/libfuzzer_libmozjpeg/hook_allocs.c](https://github.com/AFLplusplus/LibAFL/pull/2670#diff-483ae119d23d9fff6335ecf88a136032ee59ef8badebde7db67eecf42a09c340), but the CI has different ideas. Can we fix this to a point where the formatters on all systems have the same opinion?

domenukk commented:
> Also: I ran `scripts/fmt_all.sh` on MacOS. It suggested the changes in [fuzzers/inprocess/libfuzzer_libmozjpeg/hook_allocs.c](https://github.com/AFLplusplus/LibAFL/pull/2670#diff-483ae119d23d9fff6335ecf88a136032ee59ef8badebde7db67eecf42a09c340), but the CI has different ideas. Can we fix this to a point where the formatters on all systems have the same opinion?

You probably have a different version of llvm?

riesentoaster commented:
> You probably have a different version of llvm?

I do. And missed the warning. So I improved the warning.

Before:

<img width="850" alt="image" src="https://github.com/user-attachments/assets/1b122621-1ffc-4341-928a-0f1b90bb5597">

After:

<img width="943" alt="image" src="https://github.com/user-attachments/assets/813bf9fb-d29d-45e9-824a-68b7c57a29b2">

riesentoaster commented:
I've also just noticed that `black` (the python formatter) doesn't run in the CI, so I added a step to install it and changed the format script to fail if there are any changes.

domenukk commented:
We could also consider upgrading the formatter to 19 in CI, what do you think @tokatoka

tokatoka commented:
yeah it's good

riesentoaster commented:
I'm not sure what broke in the QEMU example, I don't think it has anything to do with this PR though:

```
  cargo:warning=Qemu not found, cloning with git (b01a0bc334cf11bfc5e8f121d9520ef7f47dbcd1)...
  Initialized empty Git repository in /__w/LibAFL/LibAFL/target/debug/qemu-libafl-bridge/.git/

  --- stderr
  hint: Using 'master' as the name for the initial branch. This default branch name
  hint: is subject to change. To configure the initial branch name to use in all
  hint: of your new repositories, which will suppress this warning, call:
  hint: 
  hint: 	git config --global init.defaultBranch <name>
  hint: 
  hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
  hint: 'development'. The just-created branch can be renamed via this command:
  hint: 
  hint: 	git branch -m <name>
  error: RPC failed; curl 92 HTTP/2 stream 0 was not closed cleanly: CANCEL (err 8)
  error: 18763 bytes of body are still expected
  fetch-pack: unexpected disconnect while reading sideband packet
  fatal: early EOF
  fatal: fetch-pack: invalid index-pack output
  thread 'main' panicked at /__w/LibAFL/LibAFL/libafl_qemu/libafl_qemu_build/src/build.rs:338:13:
  assertion failed: Command::new("git").current_dir(&qemu_path).arg("fetch").arg("--depth").arg("1").arg("origin").arg(QEMU_REVISION).status().unwrap().success()
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

domenukk commented:
Sounds like a network error

domenukk commented on libafl/src/events/launcher.rs:
```diff
@@ -234,66 +228,63 @@ where
+                            let builder = builder.time_ref(self.time_ref.clone());
+                            let (state, mgr) = builder.build().launch()?;
+
+                            return (self.run_client.take().unwrap())(state, mgr, *bind_to);
```
Only issue I see is that this may break clients that assume there's at most one `bind_to` core_id. They may, for example, create a directory based off of the core_id.
We probably want to add the overcommit information somewhere...
Either in the state, in an extra param, or even add the overcommit information into the core id somehow

domenukk replied:
We don't really have access to the client id at this point I think. That's mainly an LLMP thing right now. CoreId is still needed, too, since some fuzzers decide what cores to run ASan on according to the ID, for eaxmple...

domenukk replied:
For now we can just add metadata to the State for example, or see if we can easily add the info to CoreId

riesentoaster replied:
Ah, I didn't know there was a client id already. I was thinking of creating a new one, that was essentially just a u64, and maybe with some additional data (such as the core id).

domenukk replied:
so yeah store the overcommit in metadata or hack it into the ClientId

domenukk replied:
Also, please add a test for overcommit / use it on a fuzzer / ?

domenukk commented:
I think I'll merge this for now. It's not really useful yet but it doesn't have drawbacks either - and gets us the CI improvements. Feel free to follow up with PRs to make this usable in practice :)
